### PR TITLE
Speed up Daml repl startup

### PR DESCRIPTION
**Pull request by cocreature**
Daml repl startup is stupidly slow to the point where starting up and
executing 1+1 takes ~7s (piped to stdin so no typing included).

This PR speeds this up to ~4.5s. In addition to that, we also display
the repl process sooner and complete the last bit of setup while the
user is typing.

Specifically, we do the following:

1. Startup the JVM process in parallel with initializing our package
   db and only block on it once it’s started (and remove a stupid 1s
   sleep which doesn’t actually matter).
2. Load all packages in one go instead of one by one. This (roughly)
   matches how the script service works.
3. Only block on packages being loaded once we start running the
   script. Users can type before and we even typecheck their stuff
   before.

We could take this further and do even more in parallel with users
typing but this seems to strike a good balance between implementation
complexity, UX (if startup actually fails you get somewhat sensible
errors) & performance.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by akrmn — APPROVED** [review]
Looks good, thank you!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 1]
```diff
@@ -322,18 +307,16 @@ parseReplInput input dflags =
 -- | Load all packages in the given session.
 --
 -- Returns the list of modules in the specified import packages.
```
Comment (akrmn) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:309]:
```suggestion
-- Returns the list of modules in the specified import packages and an action that waits for them to be loaded by the repl client.
```

File: compiler/repl-service/client/src/DA/Daml/LF/ReplClient.hs [hunk 2]
```diff
@@ -113,21 +120,29 @@ withReplClient opts@Options{..} f = withTempFile $ \portFile -> do
...
-          replClient <- Grpc.replServiceClient client
-          f Handle
-              { hClient = replClient
-              , hOptions = opts
-              } stdout ph
-
-loadPackage :: Handle -> BS.ByteString -> IO (Either BackendError ())
-loadPackage Handle{..} package = do
+      clientBarrier <- newBarrier
+      terminateBarrier <- newBarrier
```
Comment (sofiafaro-da) [inline compiler/repl-service/client/src/DA/Daml/LF/ReplClient.hs:124]: I find it a bit confusing that `terminateBarrier` doesn't have anything to do with `hTerminate`. Maybe you could rename `terminateBarrier`, or add a comment?

Comment (sofiafaro-da) [inline compiler/repl-service/client/src/DA/Daml/LF/ReplClient.hs:124]: I find it a bit confusing that `terminateBarrier` doesn't have anything to do with `hTerminate`. Maybe you could rename `terminateBarrier`, or add a comment?

Comment (cocreature) [reply compiler/repl-service/client/src/DA/Daml/LF/ReplClient.hs]: renamed terminateBarrier to exitBarrier which is probably not much clearer but I also added a comment which hopefully is cleraer.

**Pull request merged**